File: Problems 0000-0099/03. Longest Substring Without Repeating Characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        charMap = {}
        maxLen = 0
        start = 0
        for end in range(len(s)):
            if s[end] not in charMap or charMap[s[end]] < start:
                maxLen = max(maxLen, end - start + 1)
            else:
                start = charMap[s[end]] + 1
            charMap[s[end]] = end
        return maxLen

    # def lengthOfLongestSubstring(self, s: str) -> int:
    #     seen = set()
    #     maxLen = 0
    #     start = 0
    #     for end in range(len(s)):
    #         if s[end] not in seen:
    #             maxLen = max(maxLen, end - start + 1)
    #         else:
    #             while s[end] in seen:
    #                 seen.remove(s[start])
    #                 start += 1
    #         seen.add(s[end])
    #     return maxLen

File: Problems 0000-0099/03. Longest Substring Without Repeating Characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        // Most recent index of char
        unordered_map<char, int> charMap;
        int l = 0;
        int res = 0;
        for (int r = 0; r < s.size(); r++) {

            if (charMap.count(s[r]) == 0 || charMap[s[r]] < l) {
                charMap[s[r]] = r;
                res = max(res, r - l + 1);
            } else {
                l = charMap[s[r]] + 1;
                charMap[s[r]] = r;
            }
        }
        return res;
    }
};